test/e2e/upgrade_test.go

```diff
@@ -31,15 +31,70 @@ const (
+func getNodeAffinities() ([]*v1.Affinity, error) {
+	var affinities []*v1.Affinity
+	if gc.numberOfMachineNodes >= 1 {
+		affinityNoByoh, err := getNodeAffinityForLabel(v1.NodeSelectorOpDoesNotExist, controllers.BYOHLabel)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating node affinity without BYOH label")
+		}
+		affinities = append(affinities, affinityNoByoh)
+	}
+	if gc.numberOfBYOHNodes >= 1 {
+		affinityByoh, err := getNodeAffinityForLabel(v1.NodeSelectorOpIn, controllers.BYOHLabel, "true")
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating node affinity for BYOH label")
+		}
+		affinities = append(affinities, affinityByoh)
+	}
+	return affinities, nil
+}
+
+// getLinuxJobCommand creates a Job to curl Windows webserver every 5 seconds.
+// If the webserver is not accessible for few minutes Curl Windows webserver
+// exit the pod with an error.
+func getLinuxJobCommand(clusterIP string) []string {
+	return []string{
+		"bash",
+		"-c",
+		" while true;" +
+			" do curl " + clusterIP + ";" +
+			" if [ $? != 0 ]; then" +
+			" sleep 60;" +
+			" curl " + clusterIP + " || exit 1;" +
+			" fi; " +
+			"sleep 5; " +
+			"done"}
+}
+
 // upgradeTestSuite tests behaviour of the operator when an upgrade takes place.
 func upgradeTestSuite(t *testing.T) {
 	testCtx, err := NewTestContext()
 	require.NoError(t, err)
 
-	// test if Windows workloads are running by creating a Job that curls the workloads continuously.
-	testerJob, err := testCtx.deployWindowsWorkloadAndTester()
-	require.NoError(t, err, "error testing Windows workloads")
-	defer testCtx.deleteJob(testerJob.Name)
+	// generate node affinities
+	affinities, err := getNodeAffinities()
+	require.NoError(t, err, "error getting node affinity")
+	// loop affinities
+	for _, affinity := range affinities {
+		// test if Windows workloads are running by creating a Job that curls the workloads continuously.
+		deployment, err := testCtx.deployWindowsWebServer("win-webserver", affinity)
+		require.NoErrorf(t, err, "error creating Windows Webserver deployment for upgrade test")
+
+		// create a clusterIP service which can be used to reach the Windows webserver
+		intermediarySVC, err := testCtx.createService(deployment.Name, v1.ServiceTypeClusterIP, *deployment.Spec.Selector)
+		require.NoErrorf(t, err, "error creating service for deployment %s", deployment.Name)
+
+		// create a Job object that continuously curls the webserver every 5 seconds.
+		testerJob, err := testCtx.createLinuxJob(windowsWorkloadTesterJob, getLinuxJobCommand(intermediarySVC.Spec.ClusterIP))
+		require.NoErrorf(t, err, "error creating linux job %s", windowsWorkloadTesterJob)
+
+		// stack the deferred calls in reverse to ensure the right execution order (LIFO)
+		defer testCtx.deleteJob(testerJob.Name)
+		defer testCtx.deleteService(intermediarySVC.Name)
+		defer testCtx.deleteDeployment(deployment.Name)
+	}
```

Comment:
The existing methodology for checking that the workloads did not fail is flawed:
https://github.com/openshift/windows-machine-config-operator/blob/f24ca6c568285cdbd2f0e146a70a64adadb96071/test/e2e/upgrade_test.go#L71-L74

Its just checking that the number of found failed pods is 0. But if the label selector isn't actually picking up pods from the jobs being created that number will always be 0. I suspect that that is what is currently occurring.

To make things clear with an example:
Lets say you are creating 2 jobs `windows-workload-tester-job-one` and `windows-workload-tester-job-two`
The selector is looking for failed pods with the label: `job-name=windows-workload-tester-job`. It won't find any failed pods because it wouldnt find any pods with that label at all. It needs to be doing two seperate calls to find failed pods with the label `job-name=windows-workload-tester-job-one` and `job-name=windows-workload-tester-job-two`

Instead you could have a slice which keeps track of jobs
```
var workloadTesterJobs []*batch.jobs
```
add the jobs to it as you create them, and move the check for failed pods to this function, where you can use the generated job name `job.GetName()` to search for the pods. 
But you cant just search for failed pods, you need to list all the pods and ensure there is at least one pod, and then check if any of them are in the failed state.
```
require.NotEmpty(t, pods)
// then assert status.Phase != true
```

Reply:
Created [WINC-673](https://issues.redhat.com/browse/WINC-673)